### PR TITLE
stats: Close statsExporter on close to ensure flush of buffer

### PR DESCRIPTION
**Pull request by asad-awadia**
Fixes https://github.com/DataDog/opencensus-go-exporter-datadog/issues/53

### CODE REVIEW

File: stats_test.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ func testExporter(opts Options) (*testStatsExporter, error) {
 	}
 	view.RegisterExporter(e)
 	view.SetReportingPeriod(time.Millisecond)
+	defer e.statsExporter.stop()
```
Comment (gbbr) [inline stats_test.go]: I feel like this line belongs above line 36, after the error check.

**Comment by asad-awadia** [comment]
@gbbr addressed

**Comment by asad-awadia** [comment]
hmm that seemed to have messsed up the tests

File: stats_test.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ func testExporter(opts Options) (*testStatsExporter, error) {
 	if err != nil {
```
Comment (gbbr) [inline stats_test.go]: You could add another option to fix the test noise, e.g.
```go
if opts.OnError == nil {
    opts.OnError = func(_ error) {}
}
```
That should do it, right?

**Comment by asad-awadia** [comment]
```
2019/10/11 10:00:58 Failed to export to Datadog: write udp 127.0.0.1:53386->127.0.0.1:8125: use of closed network connection
2019/10/11 10:00:58 Failed to export to Datadog: write udp 127.0.0.1:50815->127.0.0.1:8125: use of closed network connection
2019/10/11 10:00:58 Failed to export to Datadog: write udp 127.0.0.1:58252->127.0.0.1:8125: use of closed network connection
2019/10/11 10:00:58 Failed to export to Datadog: write udp 127.0.0.1:63606->127.0.0.1:8125: use of closed network connection
```
Getting this when running `go test` not sure if that's ok to ignore?

File: stats.go [hunk 3]
```diff
@@ -107,3 +107,8 @@ func (s *statsExporter) submitMetric(v *view.View, row *view.Row, metricName str
 		return fmt.Errorf("aggregation %T is not supported", v.Aggregation)
 	}
 }
+
+func (s *statsExporter) stop() {
+	err := s.client.Close()
```
Comment (gbbr) [inline stats.go]: I missed this, but here we have to do:
```go
if err := s.client.Close(); err != nil {
    s.opts.onError(err)
}
```

**Comment by gbbr** [comment]
> Getting this when running go test not sure if that's ok to ignore?

It should be fine. The noise will be reduced once you add my suggested change.

@AlexandreYang can you please confirm about https://github.com/DataDog/opencensus-go-exporter-datadog/pull/54#issuecomment-541077724? I think we don't need a running agent for these tests, thus the error.

File: stats_test.go (hunk 2 above)
Comment (asad-awadia) [reply stats_test.go]: yes that worked thanks!!

File: stats.go (hunk 3 above)
Comment (asad-awadia) [reply stats.go]: Yes - looks much cleaner!

**Comment by asad-awadia** [comment]
Adding your change has fixed the test noise

**Comment by asad-awadia** [comment]
@AlexandreYang Good to go?

File: stats.go [hunk 4]
```diff
@@ -107,3 +107,9 @@ func (s *statsExporter) submitMetric(v *view.View, row *view.Row, metricName str
 		return fmt.Errorf("aggregation %T is not supported", v.Aggregation)
 	}
 }
+
+func (s *statsExporter) stop() {
+	if err := s.client.Close(); err != nil {
+		s.opts.onError(err)
+	}
+}
```
Comment (AlexandreYang) [inline stats.go]: Could you add a simple test case for this function `stop()` ?

Just a simple test to assert that `onError` is called when `Close` returns an error.

Comment (asad-awadia) [reply stats.go]: How? I don't know if there is a mocking setup in the tests?

Comment (AlexandreYang) [reply stats.go]: Something similar to this test that test errors should be ok: https://github.com/DataDog/opencensus-go-exporter-datadog/blob/188ce88f853d1bf5f0ab23ed0e5d1b3870518a49/stats_test.go#L77-L104

**Comment by asad-awadia** [comment]
@AlexandreYang how does this look?

cc @gbbr Tests are passing for me locally? Can you advise why the CI check is failing

**Comment by AlexandreYang** [comment]
Tests are failing due to changes in `datadog-go/statsd`.

I created an issue here, we should probably pin the dependencies: https://github.com/DataDog/opencensus-go-exporter-datadog/issues/56

Those two PR should fix the falling tests:
- https://github.com/DataDog/opencensus-go-exporter-datadog/pull/55
- https://github.com/DataDog/opencensus-go-exporter-datadog/pull/57

cc @gbbr

**Comment by AlexandreYang** [comment]
@asad-awadia The fix have been merged. You probably need to rebase to make the CI pass. Thx.

**Comment by asad-awadia** [comment]
@AlexandreYang I think I messed up my rebase :( sigh

are you going to squash merge? Or should I squash all the commits into one commit and force push?

**Comment by gbbr** [comment]
@asad-awadia it's fine, we squash it when merging. Thanks for the work!

**Pull request merged**